### ISSUE TITLE
As a developer, I want to identify and implement a solution for evaluating any number of model scenarios in a single evaluation run

### ISSUE THREAD

**Issue by epag** [NOAA-OWP/wres#122]
---

Author Name: **Hank** (Hank)
Original Redmine Issue: 119109, https://vlab.noaa.gov/redmine/issues/119109
Original Date: 2023-08-04

---

See #118360 for the requirement being addressed and #115608 for the use case with slide examples.

This ticket will touch on various aspects of the code, so I'm not sure of the category. Setting it as baselines for now.

This ticket can be resolved once we have designed a solution and implemented it. Leaving as normal priority and in the backlog pending prioritization of the NWM team requirements for use case #115608.

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **Hank** (Hank)
Original Date: 2023-08-04T13:49:20Z

---

I'm guessing this means multiple @predicted@ entries, one per model evaluated. I'm not sure what the impact will be on @baseline@. Is there only one? Or are the different models each serving as a baseline for comparison with the other models?  

I'm going to estimate 128 staff hours for implementation. My only justification for that is that it "feels" like a quick ticket to address than the mapping ticket, #119105, and I estimated 256 for that.  

Thanks,

Hank

**Comment by epag** [comment]
---

Original Redmine Comment
Author Name: **James** (James)
Original Date: 2023-08-22T12:51:48Z

---

Agree with this estimate, for a first cut. 

However, you cannot have multiple @predicted@ or @baseline@ or, more generally, multiple instances of the same declaration at the same level in yaml. Something like this should be possible, though:

```<code class="yaml">
observed: 
  sources: data/DRRC2QINE.xml
  variable: QINE
predicted:
  - label: HEFS
    sources: data/drrc2ForecastsOneMonthHEFS/
    variable: SQIN
  - label: ESP
    sources: data/drrc2ForecastsOneMonthESP/
    variable: SQIN
  - label: MMEFS
    sources: data/drrc2ForecastsOneMonthMMEFS/
    variable: SQIN
</code>
```

This would address the simple case where there is one @observed@ and up to N @predicted@.

For the more complex case involving explicit pairings of @observed@/@predicted@/@baseline@, we'd probably need a new aggregator (edit: *not* part of the first cut):

```<code class="yaml">
observed: 
  sources: data/DRRC2QINE.xml
  variable: QINE
predicted:
  sources: data/drrc2ForecastsOneMonthHEFS/
  variable: SQIN
  label: HEFS
multivariable:
  - observed: 
      sources: data/DRRC2QINE.xml
      variable: QINE
    predicted:
      sources: data/drrc2ForecastsOneMonthESP/
      variable: SQIN
      label: ESP 
  - observed: 
      sources: data/DRRC2QINE.xml
      variable: QINE
    predicted:
      sources: data/drrc2ForecastsOneMonthMMEFS/
      variable: SQIN
      label: MMEFS
</code>
```

In other words, we extend the existing, simple case of one pairing with other pairings under @multivariable@.

This way, the simple thing remains simple (at least one @observed@ and @predicted@ per evaluation), but the more complex thing is possible, an extension.

**Comment by james-d-brown** [comment]
We had a user enquire about this capability in a Redmine ticket (159779).

There is an important distinction, if/when we get to this ticket:

1. Supporting a separate evaluation of up to M models within the same evaluation workflow; and
2. Supporting a multivariate evaluation, with associated multivariate statistics for up to M models together

A good example of the latter is partial correlation or partial distance correlation or conditional mutual information where the goal is to isolate the unique contribution of one model among several (that contain shared information). However, that is an extension of the first problem. Whether both are part of this ticket or two separate tickets would be TBD at implementation time, but the latter would minimally require some additional metrics to make it useful, such as the partial correlation coefficient. Both may require an improved internal plotting capability to make the plots less busy or simpler to isolate interesting features (while sticking to the original goal of making WRES a computational engine, not a visualization engine).

At the time of writing, it is unclear whether the user was looking primarily for (1) or (2) or both, but I would guess (1).

**Comment by james-d-brown** [comment]
User clarified that it is mainly the convenience and time saving aspect for now (simpler scenario, multiple single model evaluations together), i.e., you have less configuration, fewer evaluation projects to manage and you save some time loading the shared data (e.g., obs) only once too. Nominating a milestone to keep it on our radar.
